akka-http-core/src/main/scala/akka/http/javadsl/ConnectHttp.scala

```diff
@@ -29,59 +29,89 @@ object ConnectHttp {
 
   // TODO may be optimised a bit to avoid parsing the Uri entirely for the known port cases
 
-  /** Extracts host data from given Uri. */
-  def toHost(uriHost: Uri): ConnectHttp = {
-    val s = uriHost.scheme.toLowerCase(Locale.ROOT)
-    if (s == "https") new ConnectHttpsImpl(uriHost.host.address, effectivePort(s, uriHost.port))
-    else new ConnectHttpImpl(uriHost.host.address, effectivePort(s, uriHost.port))
-  }
+  /** Extracts HTTP or HTTPS connection data from given Uri. */
+  def toHost(uriHost: Uri): ConnectHttp =
+    toHost(uriHost, uriHost.port)
 
-  def toHost(host: String): ConnectHttp = {
-    if (isHttpOrHttps(host)) toHost(Uri.create(host))
-    else toHost(Uri.create(s"http://$host"))
-  }
+  /**
+   * Extract HTTP or HTTPS connection data from given host.
+   *
+   * The host string may contain a URI or a <host>:<port> pair.
+   */
+  def toHost(host: String): ConnectHttp =
+    toHost(createUriWithScheme("http", host))
 
+  /**
+   * Extracts HTTP or HTTPS connection data from given host and port.
+   *
+   * The host string may contain a URI or a <host>:<port> pair. In both cases the
+   * port is ignored.
+   *
+   * If the given port is 0, a new local port will be assigned by the operating system,
+   * which can then be retrieved by the materialized [[akka.http.javadsl.Http.ServerBinding]].
+   */
   def toHost(host: String, port: Int): ConnectHttp = {
-    require(port > 0, "port must be > 0")
-    val start = if (isHttpOrHttps(host)) host else s"http://$host"
-    toHost(Uri.create(start).port(port))
+    require(port >= 0, "port must be >= 0")
+    toHost(createUriWithScheme("http", host), port)
+  }
+
+  private def toHost(uriHost: Uri, port: Int): ConnectHttp = {
+    val s = uriHost.scheme.toLowerCase(Locale.ROOT)
+    if (s == "https") new ConnectHttpsImpl(uriHost.host.address, effectivePort(s, port))
+    else new ConnectHttpImpl(uriHost.host.address, effectivePort(s, port))
   }
 
   /**
-   * Extracts host data from given Uri.
-   * Forces an HTTPS connection to the given host, using the default HTTPS context and default port.
+   * Extracts HTTPS connection data from given host and port.
```

Comment:
Ok, sounds good. Let's clean up later on more completely.
That API came into being to avoid exponential explosion of method overloads we had to provide for bind() and all of its versions btw.